### PR TITLE
FIX-#4394: Fix issue with multiindex metadata desync

### LINKED ISSUE #4394: MultiIndex + reset_index causes issues with groupby
This one was a little complex to reproduce, but ultimately it is an issue with multiindex + reset_index + groupby, specifically when the grouping column is in the first partition. This is caused by the deferral of metadata synchronization. It should be an easy fix, the hard part was reproducing it.

```python
import modin.pandas as pd

df = # any df
df.index = pd.MultiIndex.from_tuples([(i // 10, i // 5, i) for i in range(len(df))])  # any multiindex is fine
df2 = df.reset_index(drop=False)
df.groupby(df.columns[:2]).sum()  # Error
```

The error occurs in partitions due to trying to concatenate the values in the first partition (without the multiindex) and any other partition (with a multiindex). It has a multiindex because of the metadata propagation that happens here: https://github.com/modin-project/modin/blob/c46fb05221958e6100feea597a8aec02aa184775/modin/core/dataframe/pandas/dataframe/dataframe.py#L776

but not here: https://github.com/modin-project/modin/blob/c46fb05221958e6100feea597a8aec02aa184775/modin/core/dataframe/pandas/dataframe/dataframe.py#L2597

We just need to make sure the metadata is consistent between partitions.

### PR DESCRIPTION
**Pull request by devin-petersohn**
Resolves #4394

Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4394 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Review by mvashishtha — DISMISSED** [review]

**Comment by YarShev** [comment]
@devin-petersohn, please rebase on current master and fix conflicts.

**Comment by devin-petersohn** [comment]
Thanks @YarShev for the reminder 😄. Rebased and ready for review again!

File: modin/pandas/test/dataframe/test_indexing.py [hunk 1]
```diff
@@ -1033,6 +1033,24 @@ def test_reset_index(data):
     df_equals(modin_df_cp, pd_df_cp)
 
 
+@pytest.mark.parametrize("data", test_data_values, ids=test_data_keys)
+def test_reset_index_multiindex_groupby(data):
+    # GH#4394
+    modin_df = pd.DataFrame(data)
+    pandas_df = pandas.DataFrame(data)
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_indexing.py:1040]:
```suggestion
    modin_df, pandas_df = create_test_dfs(data)
```

File: modin/pandas/test/dataframe/test_indexing.py [hunk 2]
```diff
@@ -1033,6 +1033,24 @@ def test_reset_index(data):
...
+    modin_df.index = pd.MultiIndex.from_tuples(
+        [(i // 10, i // 5, i) for i in range(len(modin_df))]
+    )
+    pandas_df.index = pandas.MultiIndex.from_tuples(
+        [(i // 10, i // 5, i) for i in range(len(pandas_df))]
+    )
+    df_equals(
+        modin_df.reset_index().groupby(list(modin_df.columns[:2])).count(),
+        pandas_df.reset_index().groupby(list(pandas_df.columns[:2])).count(),
+    )
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_indexing.py:1051]:
```suggestion
    eval_general(
        modin_df,
        pandas_df,
        lambda df: df.reset_index().groupby(list(df.columns[:2])).count()
    )
```

File: modin/pandas/test/dataframe/test_indexing.py [hunk 3]
```diff
@@ -1033,6 +1033,24 @@ def test_reset_index(data):
     df_equals(modin_df_cp, pd_df_cp)
 
 
+@pytest.mark.parametrize("data", test_data_values, ids=test_data_keys)
+def test_reset_index_multiindex_groupby(data):
+    # GH#4394
+    modin_df = pd.DataFrame(data)
+    pandas_df = pandas.DataFrame(data)
+
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_indexing.py:1041]:
```suggestion
```

**Comment by YarShev** [comment]
@devin-petersohn, lint (black) failed.

**Review by YarShev — APPROVED** [review]
LGTM!

**Pull request merged**